### PR TITLE
Braces around subobject of initialization warning fix

### PR DESCRIPTION
**Pull request by ghost**
When compiling for Mono, Apple's compiler gives a warning saying to put braces around the 0 as it is a subobject. To silence the compiler, I put braces around the 0. Simple fix.

### CODE REVIEW

**Comment by jaykrell** [comment]
This is a silly warning imho.
You can disable it.
Or use C++ `{ }`.
Or memset.

**Comment by pi1024e** [comment]
can you please elaborate on "C++ { }"?

**Comment by jaykrell** [comment]
In C++ the idiom is:

```
type name = { };
```

works for all types, idiomatic, no warning.
Mono is valid C++ and maybe some day could be compiled as such.

In C, `type name = { 0 };` is the next best thing.
It is generic. The compiler should not warn.
The only problem is sometimes unions are not fully iniitlalized.
So maybe memset?

**Review by lambdageek — APPROVED** [review]

**Comment by lambdageek** [comment]
There's apparently a number of bugs in the GCC and Clang bug trackers about `-Wmissing-braces` incorrectly warning about `= { 0 }` in C.  For example: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80454 or its predecessor https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53119.  There are similar bugs against clang.

There are a bunch of workarounds that will make the warning go away.
(For example, declaring `LifoSemaphoreWaitEntry` like this:

```
struct _LifoSemaphoreWaitEntry {
	int signaled;                                /* swapped order of signaled and condition */
	MonoCoopCond condition;
	LifoSemaphoreWaitEntry *previous;
	LifoSemaphoreWaitEntry *next;
};
```

This proposed workaround of using `= { { 0 } }` is fine, too.  The remaining members will be implicitly initialized in the same manner as statics (ie with zero).

**Comment by CoffeeFlux** [comment]
Any of the proposed workarounds are fine by me. Just noticed the warning myself and I was about to PR this, so thanks!

**Comment by pi1024e** [comment]
Maybe { } looks better instead of { {0} }

**Comment by CoffeeFlux** [comment]
In C, that's only an extension and may not compile everywhere (including MSVC?). We are not currently a C++-only codebase, so the `{ {0} }` or struct reordering seem like better solutions to me.

**Comment by pi1024e** [comment]
Done.

**Comment by pi1024e** [comment]
Is this good to merge or no?

**Comment by jaykrell** [comment]
It is an extension in C for the most part.
What is an extension is MSVC's guaranteeing to zero the entire thing if there is a union involved.
But, one shouldn't really have to know if there is a union.
memset is sadly the only fully portable way for C and C++, though it can break C++ (vtable).
memset has the advantage of getting the padding which can be important for security (otherwise padding has been known to leak data, many incidents).

**Comment by jaykrell** [comment]
Also the int should probably go last, to minimize padding for alignment.

**Comment by jaykrell** [comment]
But the pointers can go first, also solving the warning.

**Comment by pi1024e** [comment]
I did that I thought?

**Comment by pi1024e** [comment]
Done!

**Comment by akoeplinger** [comment]
It looks like this doesn't compile in C++ mode over in mono/mono:

```
16:59:52 lifo-semaphore.c: In function 'int32_t mono_lifo_semaphore_timed_wait(LifoSemaphore*, int32_t)':
16:59:52 
lifo-semaphore.c:26:42: error: braces around scalar initializer for type 'LifoSemaphoreWaitEntry* {aka _LifoSemaphoreWaitEntry*}'
16:59:52   LifoSemaphoreWaitEntry wait_entry = {{0}};
16:59:52                                           ^
16:59:52 Makefile:1534: recipe for target 'libmonoutils_la-lifo-semaphore.lo' failed
16:59:52 
make[5]: *** [libmonoutils_la-lifo-semaphore.lo] Error 1
```

File: src/mono/mono/utils/lifo-semaphore.c [hunk 1]
```diff
@@ -23,7 +23,7 @@ mono_lifo_semaphore_delete (LifoSemaphore *semaphore)
 int32_t
 mono_lifo_semaphore_timed_wait (LifoSemaphore *semaphore, int32_t timeout_ms)
 {
-	LifoSemaphoreWaitEntry wait_entry = { 0 };
+	LifoSemaphoreWaitEntry wait_entry = {{0}};
```
Comment (CoffeeFlux) [inline src/mono/mono/utils/lifo-semaphore.c]: If we're changing the struct order, this change is unneeded (and will cause C++ builds to fail).

Just remove this from the diff and I'll approve/merge. Sorry for all the trouble.

**Comment by pi1024e** [comment]
Done!

**Review by lambdageek — APPROVED** [review]

**Comment by pi1024e** [comment]
Update? @CoffeeFlux

**Comment by pi1024e** [comment]
Update? @CoffeeFlux

**Comment by CoffeeFlux** [comment]
Sorry, I was on vacation yesterday. Looks fine, just have to wait for CI both here and in mono/mono.

**Comment by pi1024e** [comment]
It's good now @CoffeeFlux

**Pull request merged**